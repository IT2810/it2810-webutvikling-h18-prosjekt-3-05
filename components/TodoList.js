import React, { Component } from 'react';
import {
      View,
      Text,
      StyleSheet,
      TouchableOpacity,
      Dimensions,
      TextInput } from 'react-native';
import  Icon  from 'react-native-vector-icons/Entypo'

class TodoList extends Component {
  constructor(props) {
    super(props),
    this.state = {
      isEditing: false,
      isCompleted: false,
      todoValue: props.textValue,
      deleteTodo: props.deleteTodo,
      id: props.id,
      inCompleteTodo: props.isCompleteTodo,
      completeTodo: props.completeTodo,
      updateTodo: props.updateTodo
    }
  }

/*Lets the user edit the todoItem.
This is triggered when the pencil-icon is clicked.*/
  startEditing = () => {
    this.setState({
      isEditing: true,
    });
  };

/*Triggered when the checked-icon is clicked.
Triggeres the updateTodo-function in TodoScreen.*/
  finishEditing = () => {
    const { todoValue } = this.state;
    const { id, updateTodo } = this.props;
    updateTodo(id, todoValue);
    this.setState({
      isEditing: false
    });
  };

    controlInput = textValue => {
        this.setState({ todoValue: textValue });
      };

/*Lets the user check list-items of as "completed"*/
      toggleItem = () => {
        const { isCompleted, inCompleteTodo, completeTodo, id } = this.props;
        if (isCompleted) {
          inCompleteTodo(id);
        } else {
          completeTodo(id);
        }
      };

  render() {
    const { isEditing, todoValue } = this.state;
    const { textValue, id, deleteTodo, isCompleted } = this.props;

        return (
            <View style={styles.container}>
              <View style={styles.rowContainer}>
              //TodoList items with (un)checked circles
              //Circle (show a different design depending on whether the Todo-item is completed or not)
                <TouchableOpacity onPress={this.toggleItem}>
                  <View style={[styles.circle, isCompleted ? styles.completeCircle : styles.incompleteCircle]}>
                  </View>
                </TouchableOpacity>
                /*If the pencil-icon is clicked,the isEditing is true,
                and the text-field turns into an input field.*/
                {isEditing ? (
                  <TextInput
                    value={this.state.todoValue}
                    autoFocus
                    maxLength={20}
                    maxLine={1}
                    style={[
                      styles.text,
                      styles.input,
                      isCompleted ? styles.strikeText : styles.unstrikeText
                    ]}
                    multiline={true}
                    returnKeyType={'done'}
                    onSubmitEditing={this.finishEditing}
                    onChangeText={this.controlInput}
                    />
                  ) : (
                    /*If isEditing is not true, the field is a normal text-field displaying the todoItem*/
                    <Text
                      style={[
                        styles.text,
                        isCompleted ? styles.strikeText : styles.unstrikeText
                      ]}
                      >
                      {this.state.todoValue}
                  </Text>
                )
              }
            </View>
            //Options for editing and deleting todoItems
            /*If the todoItem is being edited, a check-icon will appear
            for the user to click on when editing is finished.
            This triggeres the finishedEditing-function */
            {isEditing ? (
              <View style={styles.buttons}>
              <TouchableOpacity onPressOut={this.finishEditing}>
                <View style={styles.buttonContainer}>
                  <Icon name="check" style={styles.finishedText} > </Icon>
                </View>
              </TouchableOpacity>
              </View>
            ) : (
              /*If the user is not editing anything, he will see a pencil and an x icon,
              each triggering a different function*/
              <View style={styles.buttons}>
              <TouchableOpacity onPressOut={this.startEditing}>
                <View style={styles.buttonContainer}>
                  <Icon name="edit" style={styles.editText} > </Icon>
                </View>
              </TouchableOpacity>
              <TouchableOpacity onPressOut={() => deleteTodo(id)}>
                <View style={styles.buttonContainer}>
                  <Icon name="cross" style={styles.deleteText} > </Icon>
                </View>
              </TouchableOpacity>
              </View>
            )}
          </View>
        )
      };
}

//Styling
const { height, width } = Dimensions.get('window');

const styles = StyleSheet.create({
  container: {
    width: width - 50,
    borderBottomColor: '#bbb',
    borderBottomWidth: StyleSheet.hairlineWidth,
    flexDirection: 'row',
    alignItems: 'center'
  },
  text: {
    fontWeight: '500',
    fontSize: 18,
    marginVertical: 20,
  },
  circle: {
    width: 30,
    height: 30,
    borderRadius: 15,
    borderWidth: 3,
    marginRight: 20
  },
  completeCircle: {
    borderColor: '#bbb'
  },
  incompleteCircle: {
    borderColor: '#2e78b7'
  },
  rowContainer: {
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
  },
  buttons: {
    flexDirection: 'row',
  },
  buttonContainer: {
    marginVertical: 10,
    marginHorizontal: 10,
    alignSelf: 'flex-end'
},
finishedText: {
  color: '#33ff33',
  fontSize: 25
},
editText: {
  color: '#ff9900',
  fontSize: 20
},
deleteText: {
  color: 'red',
  fontSize: 25
},
strikeText: {
    color: '#bbb',
    textDecorationLine: 'line-through'
  },
  unstrikeText: {
    color: "#29323c"
  }
});

export default TodoList;
