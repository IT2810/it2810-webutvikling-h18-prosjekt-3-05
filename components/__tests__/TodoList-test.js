import React from 'react';
import TodoList from '../TodoList';
import renderer from 'react-test-renderer';

it('should render corrrectly', () => {
  const component = renderer.create(<TodoList id={1} />)
  const tree = renderer.create(<TodoList id={1} />).toJSON();
  expect (tree).toMatchSnapshot();
})

it('should set isEditing to true when startEditing is called', () => {
  const component = renderer.create(<TodoList id={1}/>)
  const instance = component.root.instance
  const props = component.root.props
  const state = instance.state
  const mockSetState = jest.fn()
  instance.setState = mockSetState;
  instance.startEditing()
  expect(mockSetState).toHaveBeenCalled()
})

it('should call updateTodo and set isEditing to false when finishEditing is called', () => {
  const fakeUpdate = jest.fn() //Mocks the updateTodo function
  const component = renderer.create(<TodoList id={1} updateTodo={fakeUpdate}/>)
  const instance = component.root.instance
  const props = component.root.props
  const state = instance.state

  instance.finishEditing()
  expect(fakeUpdate).toHaveBeenCalled()
  expect(instance.state.isEditing).toBe(false)
})

it('should call completeTodo-function in TodosScreen, when the item is toggled when isCompleted is false ', () => {
  const fakeComplete = jest.fn() //Mocks the completeTodo function
  const component = renderer.create(<TodoList id={1} isCompleted={false} completeTodo={fakeComplete}/>)
  const instance = component.root.instance
  const props = component.root.props
  const state = instance.state

  instance.toggleItem()
  expect(fakeComplete).toHaveBeenCalled()
})

it('should call inCompleteTodo-function in TodosScreen, when the item is toggled when isCompleted is true', () => {
  const fakeinComplete = jest.fn() //Mocks the inCompleteTodo function
  const component = renderer.create(<TodoList id={1} isCompleted={true} inCompleteTodo={fakeinComplete}/>)
  const instance = component.root.instance
  const props = component.root.props
  const state = instance.state

  instance.toggleItem()
  expect(fakeinComplete).toHaveBeenCalled()
})
