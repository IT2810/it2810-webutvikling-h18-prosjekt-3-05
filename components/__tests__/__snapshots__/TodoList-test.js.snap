// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render corrrectly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "borderBottomColor": "#bbb",
      "borderBottomWidth": 0.5,
      "flex": 1,
      "flexDirection": "row",
      "width": 700,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "row",
      }
    }
  >
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "borderRadius": 15,
              "borderWidth": 3,
              "height": 30,
              "marginLeft": 10,
              "marginRight": 20,
              "width": 30,
            },
            Object {
              "borderColor": "#6200ee",
            },
          ]
        }
      />
    </View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "width": "50%",
          },
          Object {
            "fontSize": 18,
            "fontWeight": "500",
            "marginVertical": 20,
          },
          Object {
            "color": "#29323c",
          },
        ]
      }
    />
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
        "marginLeft": 10,
      }
    }
  >
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Object {
            "alignSelf": "flex-end",
            "marginHorizontal": 10,
            "marginVertical": 10,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        />
      </View>
    </View>
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Object {
            "alignSelf": "flex-end",
            "marginHorizontal": 10,
            "marginVertical": 10,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        />
      </View>
    </View>
  </View>
</View>
`;
